### PR TITLE
Resize the reserved balloon

### PR DESCRIPTION
**Pull request by askervin**
This patch makes the built-in balloon types first-class citizens in their configurability and in CPU allocation order.

### CODE REVIEW

**Comment by askervin** [comment]
This patch still sticks with two questionable old design assumptions:
1. There is only one "default" balloon instance
2. If the default balloon has no explicit CPU-configuration, then it'll use the same CPUs as the "reserved" balloon. (This looks very suspicious now in the code and makes the logic difficult to understand.)

I'd be happy to simplify this design into one where the default balloon is just like any user-defined balloon. The only thing how it would differ from them is that if there is no user-defined "default" balloon type, then the balloons policy would append a built-in "default" balloon type that captures pods that are not assigned into any other balloon type.

What do you think, @klihub, @marquiz ,@kad? Would you agree, and would you prefer including this design change into this same PR? I think it would make this PR simpler to review.

**Comment by klihub** [comment]
> This patch still sticks with two questionable old design assumptions:
> 
> 1. There is only one "default" balloon instance
> 2. If the default balloon has no explicit CPU-configuration, then it'll use the same CPUs as the "reserved" balloon. (This looks very suspicious now in the code and makes the logic difficult to understand.)
> 
> I'd be happy to simplify this design into one where the default balloon is just like any user-defined balloon. The only thing how it would differ from them is that if there is no user-defined "default" balloon type, then the balloons policy would append a built-in "default" balloon type that captures pods that are not assigned into any other balloon type.
> 
> What do you think, @klihub, @marquiz ,@kad? Would you agree, and would you prefer including this design change into this same PR? I think it would make this PR simpler to review.

I you feel that it would be easier/the right course of action to cram that change into this same PR, then we should do that.

**Comment by askervin** [comment]
> I you feel that it would be easier/the right course of action to cram that change into this same PR, then we should do that.

Thanks for the assurance, I have now done that. There can be any number of default balloons, if user wishes so. And that's a perfectly valid wish. Default and reserved balloons no more share CPUs. And there is no more assumption that balloons[0] would be the reserved, and balloons[1] the default balloon.

File: docs/resource-policy/policy/balloons.md [hunk 1]
```diff
@@ -99,8 +101,15 @@ Balloons policy parameters:
   value set here is the default for all balloon types, but it can be
   overridden with the balloon type specific setting with the same
   name.
-- `balloonTypes` is a list of balloon type definitions. Each type can
-  be configured with the following parameters:
+- `balloonTypes` is a list of balloon type definitions. The order of
+  the types is significant in balloon CPU allocation and in choosing a
+  balloon type for a new container. The first type with the highest
+  `allocatorPriority` will get CPUs before other balloons when their
+  `minBalloons` > 0. A new container without balloon annotations will
```
Comment (marquiz) [inline docs/resource-policy/policy/balloons.md:108]: Just a note that should we come up with a term for these "minBalloons > 0" balloons. "pre-allocated" or "pre-inflated" or "persistent" or "non-volatile" or "sustained" or smth. So we could discuss the division between the two types of balloons. WDYT?

**Review by klihub — APPROVED** [review]

**Review by marquiz — APPROVED** [review]
One note about documentation, but not a blocker (LGTM)

File: docs/resource-policy/policy/balloons.md (hunk 1 above)
Comment (klihub) [reply docs/resource-policy/policy/balloons.md:108]: Good point. Perhaps in a follow-up PR.

Comment (askervin) [reply docs/resource-policy/policy/balloons.md:108]: @marquiz, I agree.

**Pull request merged**